@inline(__always) should not imply @inlinable

slavapestov opened the pull request:
This is a request from the standard library team. They want to force inlining of functions inside the standard library without publishing their body for inlining across module boundaries.

I added `@inlinable` in a few places in the stdlib so that this PR has no functional effect (unless I missed something). An audit should be done to remove `@inlinable` from `@inline(__always)` functions where desired.

Fixes <rdar://problem/44657000>, <https://bugs.swift.org/browse/SR-3624>, <rdar://problem/31865137>.

slavapestov committed:
AST: @inline(__always) no longer implies @inlinable

Fixes <rdar://problem/44657000>.

slavapestov committed:
stdlib: Add @inlinable to @inline(__always) declarations

These should be audited since some might not actually need to be
@inlinable, but for now:

- Anything public and @inline(__always) is now also @inlinable
- Anything @usableFromInline and @inline(__always) is now @inlinable

slavapestov commented:
@swift-ci Please smoke test

slavapestov commented:
@swift-ci Please test source compatibility

slavapestov commented:
@swift-ci Please benchmark

jrose-apple reviewed (approved):
Ooops, I thought it already was this way. Thanks, Slava!

slavapestov commented:
Oops, missed a few.

slavapestov commented:
@swift-ci Please smoke test

slavapestov commented:
@swift-ci Please benchmark

slavapestov committed:
AST: Allow @inline(__always)/@inline(never) to be applied to properties and subscripts

Previously we only permitted it on the accessor itself, but there is
no reason not to allow it on the storage declaration.

Fixes <https://bugs.swift.org/browse/SR-3624> / <rdar://problem/31865137>.

slavapestov commented:
@swift-ci Please smoke test

slavapestov commented:
@swift-ci Please benchmark

slavapestov commented:
@swift-ci Please benchmark

slavapestov commented:
@swift-ci Please smoke test

milseman commented on stdlib/public/core/StringGutsRangeReplaceable.swift:
```diff
@@ -40,7 +40,7 @@ extension _StringGuts {
   }
 
-  @usableFromInline // @testable
+  @inlinable // @testable
```
I don't think we want `@inlinable` here.

milseman commented on stdlib/public/core/StringCharacterView.swift:
```diff
@@ -50,6 +50,7 @@ extension String: BidirectionalCollection {
   }
 
   /// The number of characters in a string.
+  @inlinable
```
I don't think we need this to be inlinable for performance

milseman reviewed (commented):
Great to see this!

I have two String removals: `String.count` and `_StringGuts.isUniqueNative`.

slavapestov commented:
@milseman Do you mind making the removals yourself in a follow-up PR? I wanted this to be NFC

slavapestov commented:
@swift-ci Please smoke test

slavapestov commented:
@swift-ci Please smoke test

slavapestov commented:
@swift-ci Please smoke test

milseman commented:
Removed in https://github.com/apple/swift/pull/20601
